### PR TITLE
Add module with technical indicators

### PR DESCRIPTION
**Pull request by NathanDeVries**

### CODE REVIEW

**Comment by sitmo** [comment]
These are my initial ideas, these are mostly ponderings about the purpose & design (not so much on the code)

* Accept numpy arrays as argument. Accepting & unpacking pandas types with a decorator can indeed also be nice, and we can then also accept native lists there too? Internally  would make it based on numpy arrays. Most algorithms don't need/use pandas indices and columns names in algorithms and so we shouldn't pass them / expect them to be passed. I personally prefer data in numpy objects when making models, and I wouldn't like to be forces to convert them to pandas before being able to compute indicators.

* Merge it in with samcoml-transform. We have 10+ indicator there already, we also have a design there like generic window functions, and support for transform that have different number of output columns than input columns. Maybe that design is not fully optimal yet but then it can be polished.

* because our submodules don't have much coupling, we can think of it as a standalone module (and do so for all our submodule ideas). There are many indicator libraries on pypi, why not use those? What would set it apart? The samcoml-transform came about because it has some functions that are both useful and unique like moving_quantile, moving_slope, and multiasset ideas. When Andrejs was implementing Boruta he mentioned that there was already a mature implementation available, and he had a point IMO.

If we add it, then I would suggest we commit to the goal of making it mature enough to cover all common indicators and most use cases? Also add indicators that are not available in popular libraries (and that we use ourselves) as a justification for not using something that's already available (or alternatively donate those to a popular indicator library and become a co-author there). We should also make the interface user-friendly for potential uses by adding numpy support.

If such a commitment is too much work & deviation from the actual goal (a project on top of Factset data) then we can also decide to using available 3rd party indicators for that project, or move a limited number or new indicators into an optional submodule (or separate main module?) and not fully replicates alternatives?

**Review by mjvakili — COMMENTED** [review]
Have you benchmarked the performance of pandas versus NumPy for these calculations?

File: quantfinlib/feature/indicators/__init__.py
Comment (mjvakili) [inline quantfinlib/feature/indicators/__init__.py:1]: Have you benchmarked the performance of pandas versus numpy for these specific calculations?

File: quantfinlib/feature/indicators/_base.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import pandas as pd
+from typing import Union
+
+
+def numpy_io_support(func):
+    """Decorator to allow numpy arrays to be passed as input and output."""
+    def wrapper(*args: Union[pd.Series, np.ndarray], **kwargs) -> Union[pd.Series, np.ndarray]:
+        count_np = sum(isinstance(arg, np.ndarray) for arg in args)
+        count_pd = sum(isinstance(arg, pd.Series) for arg in args)
+        if count_pd + count_np != len(args):
```
[the remark this answers is not in the document]
Comment (mjvakili) [reply quantfinlib/feature/indicators/_base.py]: @NathanDeVries I modified this, can you have a look? Now the non-array-like arguements can be passed as keyword and positional. In the new version, the array-like arguements must have the same type, either numpy or panda series, not mixed. what do you think?

[the remark this answers is not in the document]
Comment (NathanDeVries) [reply quantfinlib/feature/indicators/_base.py]: I like how you use the annotations of the input arguments for deciding what to convert. However, I noticed some things didn't work correctly and that you removed some of the unit tests. I made some changes and fixed those tests.

[the remark this answers is not in the document]
Comment (mjvakili) [reply quantfinlib/feature/indicators/_base.py]: Looks good to me!

**Pull request merged**